### PR TITLE
feat(webhook): prevent finalizer removal

### PR DESCRIPTION
**Pull request by luphaz**
## What does this PR do?

- [ ] Adds new functionality
- [x] Alters existing functionality
- [x] Fixes a bug
- [ ] Improves documentation or testing

Please briefly describe your changes as well as the motivation behind them:
- Chaos controller guarantee `disruption` is not deleted until associated chaos pods are 
- Chaos controller guarantee finalizer is added as long as a `disruption` is NOT deleted [here](https://github.com/DataDog/chaos-controller/blob/main/controllers/disruption_controller.go#L209)
- However, we did not prevent `disruption` finalizer removal AFTER a `disruption` was marked for deletion
- Hence, a user seeing a `disruption` blocked could be tempted to remove the finalizer to force/speed up the `disruption` removal, leading to all associated chaos pods being left un-supervised in the cluster in their statuses, forever
- The proposed implementation blocks such removal and list associated chaos pods waiting for pending action to enable/unblock `disruption` deletion

During testing I noticed the following:
* we were not using `-race` flag so I added it and fixed associated data race detected [here](https://github.com/DataDog/chaos-controller/pull/585/files#diff-71cea628d383a9b3f0e1a950238cd4543ec0994beb01e9201b3f87c2815fda30R143) and [here](https://github.com/DataDog/chaos-controller/pull/585/files#diff-d3922ac8e373158d56a2ce016a6bfc8f97d576f2f0149d971037685d1d4a0cb9R75)
* running e2e-tests locally failed for 1 test due to the `expiredDisruptionGCDelay` being `10m` and testing requiring full cleanup 
  * I suggested another implementation to be able to have all e2e-tests successful locally, however, the drawback is it COULD lead to local disruptions being deleted earlier than expected (as the GC delay is a configuration change requiring a new chaos controller deploy)
  * Let me know if you believe we should keep the old way of doing this

## Code Quality Checklist

- [ ] The documentation is up to date.
- [x] My code is sufficiently commented and passes continuous integration checks.
- [x] I have signed my commit (see [Contributing Docs](../CONTRIBUTING.md)).

## Testing

- [x] I leveraged continuous integration testing
    - [ ] by depending on existing `unit` tests or `end-to-end` tests.
    - [x] by adding new `unit` tests or `end-to-end` tests.
- [ ] I manually tested the following steps:
    - [x] locally.
    - [ ] as a canary deployment to a cluster.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -65,6 +68,10 @@ colima-all:
 	$(MAKE) colima-build
 	$(MAKE) colima-install
 
+colima-redo:
```
Comment (ptnapoleon) [inline Makefile:71]: What does this reset?

File: controllers/disruption_controller.go [hunk 2]
```diff
@@ -140,7 +140,7 @@ func (r *DisruptionReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	}
 
 	// check whether the object is being deleted or not
-	if !instance.ObjectMeta.DeletionTimestamp.IsZero() {
+	if !instance.DeletionTimestamp.IsZero() {
```
Comment (ptnapoleon) [inline controllers/disruption_controller.go:143]: Afaict, `instance.DeletionTimestamp` is the exact same as `instance.ObjectMeta.DeletionTimestamp`, yes?

File: api/v1beta1/disruption_webhook.go [hunk 3]
```diff
@@ -138,15 +142,45 @@ func (r *Disruption) ValidateCreate() error {
...
+
+	var err error
+
+	oldDisruption := old.(*Disruption)
+
+	// ensure finalizer removal is only allowed if no related chaos pods exists
+	// such verification is only needed if disruption is currently in removal as controller will always add finalizer to disruptions not in removal during the reconcile loop
+	// we should NOT always prevent finalizer removal because chaos controller reconcile loop will go through this mutating webhook when perfoming updates
+	// and need to be able to remove the finalizer to enable the disruption to be garbage collected on successful removal
+	if !oldDisruption.ObjectMeta.DeletionTimestamp.IsZero() && controllerutil.ContainsFinalizer(oldDisruption, chaostypes.DisruptionFinalizer) && !controllerutil.ContainsFinalizer(r, chaostypes.DisruptionFinalizer) {
```
Comment (ptnapoleon) [inline api/v1beta1/disruption_webhook.go:156]: Why use `oldDisruption.ObjectMeta.DeletionTimestamp` here instead of `oldDisruption.DeletionTimestamp` given similar changes elsewhere?

File: api/v1beta1/disruption_webhook.go [hunk 4]
```diff
@@ -138,15 +142,45 @@ func (r *Disruption) ValidateCreate() error {
...
-	logger.Debugw("validating updated disruption", "instance", r.Name, "namespace", r.Namespace)
+	logger := logger.With("instance", r.Name, "namespace", r.Namespace)
+	logger.Debugw("validating updated disruption")
+
+	var err error
+
+	oldDisruption := old.(*Disruption)
+
+	// ensure finalizer removal is only allowed if no related chaos pods exists
+	// such verification is only needed if disruption is currently in removal as controller will always add finalizer to disruptions not in removal during the reconcile loop
```
Comment (ptnapoleon) [inline api/v1beta1/disruption_webhook.go:153]: Is this resilient to a user first removing the finalizer, then deleting the disruption? Are we guaranteed the controller will reconcile between the two as a response to the update that removes the finalizer?

File: api/v1beta1/v1beta1_suite_test.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+)
+
+func TestV1Beta1(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	RunSpecs(t, "v1beta1 Suite")
+}
+
+var _ = BeforeSuite(func() {
+	By("initiliazing disruption_webhook global variables")
```
Comment (ptnapoleon) [inline api/v1beta1/v1beta1_suite_test.go:25]:
```suggestion
	By("initializing disruption_webhook global variables")
```

File: Makefile (hunk 1 above)
Comment (luphaz) [reply Makefile]: I changed the name to `colima-deploy` that is doing build/install/restart.

If we believe it's not relevant I will remove it, let me know.

File: controllers/disruption_controller.go (hunk 2 above)
Comment (ptnapoleon) [reply controllers/disruption_controller.go:143]: We don't favor one or the other, afaik. I was more curious if you knew a reason that the shorter form was preferable. If they're identical but one is more concise, I'm definitely in support of the shorter option

File: api/v1beta1/disruption_webhook.go (hunk 4 above)
Comment (ptnapoleon) [reply api/v1beta1/disruption_webhook.go]: I didn't think it through enough, I rescind my worries

File: controllers/disruption_controller.go (hunk 2 above)
Comment (luphaz) [reply controllers/disruption_controller.go:143]: Not really, I favoured it because it was the shorter option indeed.
The main difference is the short option due to go, is that we can have a meaning that changes:
* as long as there is no field in Disruption called `DeletionTimestamp` it's linking to `ObjectMeta.DeletionTimestamp`
* if at any time we have a second embedded struct containing this field, it will lead to a compile time error 
```text
ambiguous selector instance.DeletionTimestampcompiler[AmbiguousSelector](https://pkg.go.dev/golang.org/x/tools/internal/typesinternal?utm_source%3Dgopls#AmbiguousSelector)
```
* if we create the same field `DeletionTimestamp` we would have a compile error due to missing method

The only risk in using the shortest would hence be if we introduce the exact same field with the exact same type. I would say it's a rather edge case we might not be interested in preventing.

File: Makefile [hunk 6]
```diff
@@ -29,11 +32,16 @@ all: manager injector handler
...
 # Run unit tests
 test: generate manifests
-	go test $(shell go list ./... | grep -v chaos-controller/controllers) -coverprofile cover.out
+	go test -race $(shell go list ./... | grep -v chaos-controller/controllers) -coverprofile cover.out
 
 # Run e2e tests (against a real cluster)
-e2e-test: generate minikube-install
-	USE_EXISTING_CLUSTER=true go test ./controllers/... -coverprofile cover.out
+e2e-test: generate
+	$(MAKE) minikube-install EXPIRED_DISRUPTION_GC_DELAY=10s
```
Comment (ptnapoleon) [inline Makefile:39]: wow, so fancy!

**Review by ptnapoleon — APPROVED** [review]

**Review by ptnapoleon — APPROVED** [review]
